File: sdks/html5-javascript/lib/modules/Collection.js
/*
 *Licensed to the Apache Software Foundation (ASF) under one
 *or more contributor license agreements.  See the NOTICE file
 *distributed with this work for additional information
 *regarding copyright ownership.  The ASF licenses this file
 *to you under the Apache License, Version 2.0 (the
 *"License"); you may not use this file except in compliance
 *with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0

 *Unless required by applicable law or agreed to in writing,
 *software distributed under the License is distributed on an
 *"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *KIND, either express or implied.  See the License for the
 *specific language governing permissions and limitations
 *under the License.
 */


/*
 *  The Collection class models Usergrid Collections.  It essentially
 *  acts as a container for holding Entity objects, while providing
 *  additional funcitonality such as paging, and saving
 *
 *  @constructor
 *  @param {string} options - configuration object
 *  @return {Collection} collection
 */
Usergrid.Collection = function(options) {

  if (options) {
    this._client = options.client;
    this._type = options.type;
    this.qs = options.qs || {};

    //iteration
    this._list = options.list || [];
    this._iterator = options.iterator || -1; //first thing we do is increment, so set to -1

    //paging
    this._previous = options.previous || [];
    this._next = options.next || null;
    this._cursor = options.cursor || null;

    //restore entities if available
    if (options.list) {
      var count = options.list.length;
      for (var i = 0; i < count; i++) {
        //make new entity with
        var entity = this._client.restoreEntity(options.list[i]);
        this._list[i] = entity;
      }
    }
  }
  /*if (callback) {
    //populate the collection
    this.fetch(callback);
  }*/

};


/*
 *  method to determine whether or not the passed variable is a Usergrid Collection
 *
 *  @method isCollection
 *  @public
 *  @params {any} obj - any variable
 *  @return {boolean} Returns true or false
 */
Usergrid.isCollection = function(obj) {
  return (obj && obj instanceof Usergrid.Collection);
};


/*
 *  gets the data from the collection object for serialization
 *
 *  @method serialize
 *  @return {object} data
 */
Usergrid.Collection.prototype.serialize = function() {

  //pull out the state from this object and return it
  var data = {};
  data.type = this._type;
  data.qs = this.qs;
  data.iterator = this._iterator;
  data.previous = this._previous;
  data.next = this._next;
  data.cursor = this._cursor;

  this.resetEntityPointer();
  var i = 0;
  data.list = [];
  while (this.hasNextEntity()) {
    var entity = this.getNextEntity();
    data.list[i] = entity.serialize();
    i++;
  }

  data = JSON.stringify(data);
  return data;
};
//addCollection is deprecated?
/*Usergrid.Collection.prototype.addCollection = function (collectionName, options, callback) {
  self = this;
  options.client = this._client;
  var collection = new Usergrid.Collection(options, function(err, data) {
    if (typeof(callback) === 'function') {

      collection.resetEntityPointer();
      while(collection.hasNextEntity()) {
        var user = collection.getNextEntity();
        var email = user.get('email');
        var image = self._client.getDisplayImage(user.get('email'), user.get('picture'));
        user._portal_image_icon = image;
      }

      self[collectionName] = collection;
      doCallback(callback, [err, collection], self);
    }
  });
};*/

/*
 *  Populates the collection from the server
 *
 *  @method fetch
 *  @param {function} callback
 *  @return {callback} callback(err, data)
 */
Usergrid.Collection.prototype.fetch = function(callback) {
  var self = this;
  var qs = this.qs;

  //add in the cursor if one is available
  if (this._cursor) {
    qs.cursor = this._cursor;
  } else {
    delete qs.cursor;
  }
  var options = {
    method: 'GET',
    endpoint: this._type,
    qs: this.qs
  };
  this._client.request(options, function(err, response) {
    if (err && self._client.logging) {
      console.log('error getting collection');
    } else {
      //save the cursor if there is one
      self.saveCursor(response.cursor || null);
      self.resetEntityPointer();
      //save entities locally
      self._list = response.getEntities()
        .filter(function(entity) {
          return isUUID(entity.uuid);
        })
        .map(function(entity) {
          var ent = new Usergrid.Entity({
            client: self._client
          });
          ent.set(entity);
          ent.type = self._type;
          //ent._json = JSON.stringify(entity, null, 2);
          return ent;
        });
    }
    doCallback(callback, [err, response, self], self);
  });
};

/*
 *  Adds a new Entity to the collection (saves, then adds to the local object)
 *
 *  @method addNewEntity
 *  @param {object} entity
 *  @param {function} callback
 *  @return {callback} callback(err, data, entity)
 */
Usergrid.Collection.prototype.addEntity = function(entityObject, callback) {
  var self = this;
  entityObject.type = this._type;

  //create the new entity
  this._client.createEntity(entityObject, function(err, response, entity) {
    if (!err) {
      //then add the entity to the list
      self.addExistingEntity(entity);
    }
    doCallback(callback, [err, response, self], self);
  });
};

Usergrid.Collection.prototype.addExistingEntity = function(entity) {
  //entity should already exist in the db, so just add it to the list
  var count = this._list.length;
  this._list[count] = entity;
};

/*
 *  Removes the Entity from the collection, then destroys the object on the server
 *
 *  @method destroyEntity
 *  @param {object} entity
 *  @param {function} callback
 *  @return {callback} callback(err, data)
 */
Usergrid.Collection.prototype.destroyEntity = function(entity, callback) {
  var self = this;
  entity.destroy(function(err, response) {
    if (err) {
      if (self._client.logging) {
        console.log('could not destroy entity');
      }
      doCallback(callback, [err, response, self], self);
    } else {
      //destroy was good, so repopulate the collection
      self.fetch(callback);
    }
    //remove entity from the local store
    self.removeEntity(entity);
  });
};

/*
 * Filters the list of entities based on the supplied criteria function
 * works like Array.prototype.filter
 *
 *  @method getEntitiesByCriteria
 *  @param {function} criteria  A function that takes each entity as an argument and returns true or false
 *  @return {Entity[]} returns a list of entities that caused the criteria function to return true
 */
Usergrid.Collection.prototype.getEntitiesByCriteria = function(criteria) {
  return this._list.filter(criteria);
};
/*
 * Returns the first entity from the list of entities based on the supplied criteria function
 * works like Array.prototype.filter
 *
 *  @method getEntitiesByCriteria
 *  @param {function} criteria  A function that takes each entity as an argument and returns true or false
 *  @return {Entity[]} returns a list of entities that caused the criteria function to return true
 */
Usergrid.Collection.prototype.getEntityByCriteria = function(criteria) {
  return this.getEntitiesByCriteria(criteria).shift();
};
/*
 * Removed an entity from the collection without destroying it on the server
 *
 *  @method removeEntity
 *  @param {object} entity
 *  @return {Entity} returns the removed entity or undefined if it was not found
 */
Usergrid.Collection.prototype.removeEntity = function(entity) {
  var removedEntity = this.getEntityByCriteria(function(item) {
    return entity.uuid === item.get('uuid');
  });
  delete this._list[this._list.indexOf(removedEntity)];
  return removedEntity;
};

/*
 *  Looks up an Entity by UUID
 *
 *  @method getEntityByUUID
 *  @param {string} UUID
 *  @param {function} callback
 *  @return {callback} callback(err, data, entity)
 */
Usergrid.Collection.prototype.getEntityByUUID = function(uuid, callback) {
  var entity = this.getEntityByCriteria(function(item) {
    return item.get('uuid') === uuid;
  });
  if (entity) {
    doCallback(callback, [null, entity, entity], this);
  } else {
    //get the entity from the database
    var options = {
      data: {
        type: this._type,
        uuid: uuid
      },
      client: this._client
    };
    entity = new Usergrid.Entity(options);
    entity.fetch(callback);
  }
};

/*
 *  Returns the first Entity of the Entity list - does not affect the iterator
 *
 *  @method getFirstEntity
 *  @return {object} returns an entity object
 */
Usergrid.Collection.prototype.getFirstEntity = function() {
  var count = this._list.length;
  if (count > 0) {
    return this._list[0];
  }
  return null;
};

/*
 *  Returns the last Entity of the Entity list - does not affect the iterator
 *
 *  @method getLastEntity
 *  @return {object} returns an entity object
 */
Usergrid.Collection.prototype.getLastEntity = function() {
  var count = this._list.length;
  if (count > 0) {
    return this._list[count - 1];
  }
  return null;
};

/*
 *  Entity iteration -Checks to see if there is a "next" entity
 *  in the list.  The first time this method is called on an entity
 *  list, or after the resetEntityPointer method is called, it will
 *  return true referencing the first entity in the list
 *
 *  @method hasNextEntity
 *  @return {boolean} true if there is a next entity, false if not
 */
Usergrid.Collection.prototype.hasNextEntity = function() {
  var next = this._iterator + 1;
  var hasNextElement = (next >= 0 && next < this._list.length);
  if (hasNextElement) {
    return true;
  }
  return false;
};

/*
 *  Entity iteration - Gets the "next" entity in the list.  The first
 *  time this method is called on an entity list, or after the method
 *  resetEntityPointer is called, it will return the,
 *  first entity in the list
 *
 *  @method hasNextEntity
 *  @return {object} entity
 */
Usergrid.Collection.prototype.getNextEntity = function() {
  this._iterator++;
  var hasNextElement = (this._iterator >= 0 && this._iterator <= this._list.length);
  if (hasNextElement) {
    return this._list[this._iterator];
  }
  return false;
};

/*
 *  Entity iteration - Checks to see if there is a "previous"
 *  entity in the list.
 *
 *  @method hasPrevEntity
 *  @return {boolean} true if there is a previous entity, false if not
 */
Usergrid.Collection.prototype.hasPrevEntity = function() {
  var previous = this._iterator - 1;
  var hasPreviousElement = (previous >= 0 && previous < this._list.length);
  if (hasPreviousElement) {
    return true;
  }
  return false;
};

/*
 *  Entity iteration - Gets the "previous" entity in the list.
 *
 *  @method getPrevEntity
 *  @return {object} entity
 */
Usergrid.Collection.prototype.getPrevEntity = function() {
  this._iterator--;
  var hasPreviousElement = (this._iterator >= 0 && this._iterator <= this._list.length);
  if (hasPreviousElement) {
    return this._list[this._iterator];
  }
  return false;
};

/*
 *  Entity iteration - Resets the iterator back to the beginning
 *  of the list
 *
 *  @method resetEntityPointer
 *  @return none
 */
Usergrid.Collection.prototype.resetEntityPointer = function() {
  this._iterator = -1;
};

/*
 * Method to save off the cursor just returned by the last API call
 *
 * @public
 * @method saveCursor
 * @return none
 */
Usergrid.Collection.prototype.saveCursor = function(cursor) {
  //if current cursor is different, grab it for next cursor
  if (this._next !== cursor) {
    this._next = cursor;
  }
};

/*
 * Resets the paging pointer (back to original page)
 *
 * @public
 * @method resetPaging
 * @return none
 */
Usergrid.Collection.prototype.resetPaging = function() {
  this._previous = [];
  this._next = null;
  this._cursor = null;
};

/*
 *  Paging -  checks to see if there is a next page od data
 *
 *  @method hasNextPage
 *  @return {boolean} returns true if there is a next page of data, false otherwise
 */
Usergrid.Collection.prototype.hasNextPage = function() {
  return (this._next);
};

/*
 *  Paging - advances the cursor and gets the next
 *  page of data from the API.  Stores returned entities
 *  in the Entity list.
 *
 *  @method getNextPage
 *  @param {function} callback
 *  @return {callback} callback(err, data)
 */
Usergrid.Collection.prototype.getNextPage = function(callback) {
  if (this.hasNextPage()) {
    //set the cursor to the next page of data
    this._previous.push(this._cursor);
    this._cursor = this._next;
    //empty the list
    this._list = [];
    this.fetch(callback);
  }
};

/*
 *  Paging -  checks to see if there is a previous page od data
 *
 *  @method hasPreviousPage
 *  @return {boolean} returns true if there is a previous page of data, false otherwise
 */
Usergrid.Collection.prototype.hasPreviousPage = function() {
  return (this._previous.length > 0);
};

/*
 *  Paging - reverts the cursor and gets the previous
 *  page of data from the API.  Stores returned entities
 *  in the Entity list.
 *
 *  @method getPreviousPage
 *  @param {function} callback
 *  @return {callback} callback(err, data)
 */
Usergrid.Collection.prototype.getPreviousPage = function(callback) {
  if (this.hasPreviousPage()) {
    this._next = null; //clear out next so the comparison will find the next item
    this._cursor = this._previous.pop();
    //empty the list
    this._list = [];
    this.fetch(callback);
  }
};

File: sdks/android/src/main/java/org/apache/usergrid/android/sdk/entities/Collection.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.usergrid.android.sdk.entities;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.apache.usergrid.android.sdk.UGClient;
import org.apache.usergrid.android.sdk.UGClient.Query;
import org.apache.usergrid.android.sdk.response.ApiResponse;

/**
 * Models a collection of entities as a local object. Collections
 * are the primary way that entities are organized in Usergrid.
 *
 * @see <a href="http://apigee.com/docs/app-services/content/collections">Collections documentation</a>
 */
public class Collection
{
	private UGClient client;
	private String type;
	private Map<String,Object> qs;

	private ArrayList<Entity> list;
	private int iterator;
	private ArrayList<String> previous;
	private String next;
	private String cursor;

	/**
	 * Default constructor for a Collection object.
	 *
	 * @param  UGClient  an instance of the UGClient class
	 * @param  type  the entity 'type' associated with the colleciton
	 * @param  qs  optional Map object of query parameters to apply to the collection retrieval
	 */	
	public Collection(UGClient client, String type, Map<String,Object> qs) {
	    this.client = client;
	    this.type = type;
	    
	    if( qs == null )
	    {
	    	this.qs = new HashMap<String,Object>();
	    }
	    else
	    {
	    	this.qs = qs;
	    }

	    this.list = new ArrayList<Entity>();
	    this.iterator = -1;

	    this.previous = new ArrayList<String>();
	    this.next = null;
	    this.cursor = null;

	    this.fetch();
	}

	/**
	 * Gets the entity 'type' associated with the collection.
	 *
	 * @return the collection type
	 */	
	public String getType(){
	   return this.type;
	}
	
	/**
	 * Sets the entity 'type' associated with the collection.
	 *
	 * @param  type  the collection type
	 */	
	public void setType(String type){
	   this.type = type;
	}

	/**
	 * Retrieves the current state of the collection from the server, and populates
	 * an the Collection object with the returned set of entities. Executes synchronously.
	 *
	 * @return  an ApiResponse object
	 */	
	public ApiResponse fetch() {
	    if (this.cursor != null) {
	    	this.qs.put("cursor", this.cursor);
	    }
	    
	    Query query = this.client.queryEntitiesRequest("GET", this.qs, null,
                this.client.getOrganizationId(),  this.client.getApplicationId(), this.type);
	    ApiResponse response = query.getResponse();
	    if (response.getError() != null) {
	    	this.client.writeLog("Error getting collection.");
	    } else {
	    	String theCursor = response.getCursor();
    		int count = response.getEntityCount();
    		
    		UUID nextUUID = response.getNext();
    		if( nextUUID != null ) {
    			this.next = nextUUID.toString();
    		} else {
    			this.next = null;
    		}
    		this.cursor = theCursor;

	    	this.saveCursor(theCursor);
	    	if ( count > 0 ) {
	    		this.resetEntityPointer();
	    		this.list = new ArrayList<Entity>();
	    		List<Entity> retrievedEntities = response.getEntities();
	    		
	    		for ( Entity retrievedEntity : retrievedEntities ) {
	    			if( retrievedEntity.getUuid() != null ) {
	    				retrievedEntity.setType(this.type);
	    				this.list.add(retrievedEntity);
	    			}
	    		}
	    	}
	    }
	    
	    return response;
	}

	/**
	 * Adds an entity to the Collection object.
	 *
	 * @param  entityData  a Map object of entity properties to be saved in the entity
	 * @return  an Entity object that represents the newly added entity. Must include
	 *		a 'type' property. Executes synchronously.
	 */	
	public Entity addEntity(Map<String,Object> entityData) {
		Entity entity = null;		
		ApiResponse response = this.client.createEntity(entityData);
		if( (response != null) && (response.getError() == null) && (response.getEntityCount() > 0) ) {
			entity = response.getFirstEntity();
			if (entity != null) {
				this.list.add(entity);
			}
		}
		return entity;
	}

	/**
	 * Deletes the provided entity on the server, then updates the
	 * Collection object by calling fetch(). Executes synchronously.
	 *
	 * @param entity an Entity object that contains a 'type' and 'uuid' property
	 */	
	public ApiResponse destroyEntity(Entity entity) {
		ApiResponse response = entity.destroy();
		if (response.getError() != null) {
			this.client.writeLog("Could not destroy entity.");
		} else {
			response = this.fetch();
		}
	    
		return response;
	}

	/**
	 * Retrieves an entity from the server.
	 *
	 * @param uuid the UUID of the entity to retrieve
	 * @return an ApiResponse object
	 */	
	public ApiResponse getEntityByUuid(UUID uuid) {
		Entity entity = new Entity(this.client);
	    entity.setType(this.type);
	    entity.setUuid(uuid);
	    return entity.fetch();
	}

	/**
	 * Gets the first entity in the Collection object.
	 *
	 * @return  an Entity object
	 */	
	public Entity getFirstEntity() {
		return ((this.list.size() > 0) ? this.list.get(0) : null);
	}

	/**
	 * Gets the last entity in the Collection object.
	 *
	 * @return  an Entity object
	 */	
	public Entity getLastEntity() {
		return ((this.list.size() > 0) ? this.list.get(this.list.size()-1) : null);
	}

	/**
	 * Checks if there is another entity in the Collection after the current pointer position.
	 *
	 * @return  Boolean true/false
	 */	
	public boolean hasNextEntity() {
		int next = this.iterator + 1;
		return ((next >= 0) && (next < this.list.size()));
	}

	/**
	 * Checks if there is an entity in the Collection before the current pointer position.
	 *
	 * @return  Boolean true/false
	 */	
	public boolean hasPrevEntity() {
		int prev = this.iterator - 1;
		return ((prev >= 0) && (prev < this.list.size()));
	}

	/**
	 * Checks if there is an entity in the Collection after the current
	 * pointer position, and returns it.
	 *
	 * @return an Entity object
	 */	
	public Entity getNextEntity() {
		if (this.hasNextEntity()) {
			this.iterator++;
			return this.list.get(this.iterator);
		}
		return null;
	}

	/**
	 * Checks if there is an entity in the Collection before the current
	 * pointer position, and returns it.
	 *
	 * @return an Entity object
	 */	
	public Entity getPrevEntity() {
		if (this.hasPrevEntity()) {
			this.iterator--;
			return this.list.get(this.iterator);
		}
		return null;
	}

	/**
	 * Resets the pointer to the start of the Collection.
	 */
	public void resetEntityPointer() {
		this.iterator = -1;
	}

	/**
	 * Saves a pagination cursor.	 
	 */
	public void saveCursor(String cursor) {
		this.next = cursor;
	}

	/**
	 * Clears the currently saved pagination cursor from the Collection.
	 */
	public void resetPaging() {
		this.previous.clear();
		this.next = null;
		this.cursor = null;
	}

	/**
	 * Checks if a pagination cursor for the next result set is 
	 * present in the Collection.
	 *
	 * @return Boolean true/false
	 */
	public boolean hasNextPage() {
		return this.next != null;
	}

	/**
	 * Checks if a pagination cursor for the previous result set is 
	 * present in the Collection
	 *
	 * @return  Boolean true/false
	 */
	public boolean hasPrevPage() {
		return !this.previous.isEmpty();
	}

	/**
	 * Checks if a pagination cursor for the next result set is 
	 * present in the Collection, then fetches it.
	 *
	 * @return an ApiResponse object if a cursor is present, otherwise null
	 */
	public ApiResponse getNextPage() {
		if (this.hasNextPage()) {
			this.previous.add(this.cursor);
			this.cursor = this.next;
			this.list.clear();
			return this.fetch();
		}
		  
		return null;
	}

	/**
	 * Checks if a pagination cursor for the previous result set is 
	 * present in the Collection, then fetches it.
	 *
	 * @return  an ApiResponse object if a cursor is present, otherwise null
	 */
	public ApiResponse getPrevPage() {
		if (this.hasPrevPage()) {
			this.next = null;
			int indexOfLastObject = this.previous.size() - 1;
			this.cursor = this.previous.get(indexOfLastObject);
			this.previous.remove(indexOfLastObject);
			this.list.clear();
			return this.fetch();
		}
		  
		return null;
	}

}
